pkg/trace/writer/trace.go

```diff
@@ -139,6 +145,49 @@ func (w *TraceWriter) Run() {
 	}
 }
 
+func (w *TraceWriter) runSync() {
+	defer close(w.stop)
+	defer close(w.flushChan)
+	for {
+		select {
+		case pkg := <-w.In:
+			w.addSpans(pkg)
+		case notify := <-w.flushChan:
+			w.drainAndFlush()
+			// Wait for encoding/compression to complete on each payload,
+			// and submission to senders
+			w.wg.Wait()
+			notify <- struct{}{}
+		case <-w.stop:
+			w.drainAndFlush()
+			return
+		}
+	}
+}
+
+// FlushSync blocks and sends pending payloads when syncMode is true
+func (w *TraceWriter) FlushSync() error {
+	if !w.syncMode {
+		return errors.New("FlushSync called on TraceWriter when in async mode")
+	}
+	defer w.report()
+
+	notify := make(chan struct{}, 1)
+	w.flushChan <- notify
+	<-notify
+
+	var wg sync.WaitGroup
+	for _, s := range w.senders {
+		wg.Add(1)
+		go func(s *sender) {
+			defer wg.Done()
+			s.waitForInflight()
+		}(s)
+	}
+	wg.Wait()
```

Comment:
This here is also out of place. I feel like it's a duplication. Let's factor it out into a global function and call it inside the `select` case just like I proposed for the stats writer. Maybe something like `waitForSenders`